File: Assigment 2: Deques and Randomized Queues/Python/Deque.py
"""
The `Deque` class represents a double-ended queue (deque) that allows adding and removing items
from both ends of the collection. It uses a doubly linked list implementation for efficient
operations at both ends.

Usage:
- Create a deque: `my_deque = Deque()`
- Add items to the front: `my_deque.addFirst(item)`
- Add items to the back: `my_deque.addLast(item)`
- Remove an item from the front: `my_deque.removeFirst()`
- Remove an item from the back: `my_deque.removeLast()`
- Iterate over items: `for item in my_deque.Iterator(): ...`

Note: The `Deque` class includes an inner `Node` class for representing nodes in the linked list.
"""

class Deque:
    def __init__(self):
        """
        Initializes an empty deque.
        """
        self.first = None
        self.last = None
        self.size = 0

    class Node:
        def __init__(self, val, next = None, prev = None):
            """
            Represents a node in the doubly linked list.

            Args:
                val: The value stored in the node.
                next: Reference to the next node.
                prev: Reference to the previous node.
            """
            self.val = val
            self.next = next
            self.prev = prev

    class IteratorClass:
        def __init__(self, head):
            """
            Initializes an iterator for the deque.

            Args:
                head: The starting node for iteration.
            """
            self.head = head

        def __iter__(self):
            return self

        def __next__(self):
            """
            Returns the next item during iteration.

            Raises:
                StopIteration: When there are no more items to iterate.

            Returns:
                The value of the current node.
            """
            curr = self.head
            if curr:
                item = curr.val
                self.head = curr.next
                return item
            else:
                raise StopIteration

    def isEmpty(self):
        """
        Checks if the deque is empty.

        Returns:
            True if the deque is empty, False otherwise.
        """
        return self.first == None

    def size(self):
        """
        Returns the number of items in the deque.

        Returns:
            The number of items.
        """
        return self.size

    def addFirst(self, val):
        """
        Adds an item to the front of the deque.

        Args:
            val: The item to be added.

        Raises:
            ValueError: If the input item is None.
        """
        self.size += 1
        if self.isEmpty():
            self.first = self.last = self.Node(val)
        else:
            new_node = self.Node(val, next = self.first)
            self.first.prev = new_node
            self.first = new_node

    def addLast(self, val):
        """
        Adds an item to the back of the deque.

        Args:
            val: The item to be added.

        Raises:
            ValueError: If the input item is None.
        """
        self.size += 1
        if self.isEmpty():
            self.first = self.last = self.Node(val)
        else:
            new_node = self.Node(val, prev = self.last)
            self.last.next = new_node
            self.last = new_node

    def removeFirst(self):
        """
        Removes and returns the item from the front of the deque.

        Returns:
            The removed item.

        Raises:
            IndexError: If the deque is empty.
        """
        if self.isEmpty():
            print('The list is empty')
        else:
            self.size -= 1
            item = self.first.val
            self.first = self.first.next
            if self.first == None:
                self.last = None
            else:
                self.first.prev = None
            return item

    def removeLast(self):
        """
        Removes and returns the item from the back of the deque.

        Returns:
            The removed item.

        Raises:
            IndexError: If the deque is empty.
        """
        if self.isEmpty():
            print('The list is empty')
        else:
            self.size -= 1
            item = self.last.val
            self.last = self.last.prev
            if self.last == None:
                self.first = None
            else:
                self.last.next = None
            return item

    def Iterator(self):
        """
        Returns an iterator over items in order from front to back.

        Returns:
            An iterator.
        """
        return self.IteratorClass(self.first)

    def __str__(self):
        """
        Returns a string representation of the deque.

        Returns:
            A string showing the items in the deque.
        """
        curr = self.first
        sequence = []
        while curr:
            sequence.append(str(curr.val))
            curr = curr.next
        return ' <-> '.join(sequence)

    def reversed(self):
        """
        Returns a string representation of the deque in reverse order.

        Returns:
            A string showing the items in reverse order.
        """
        curr = self.last
        sequence = []
        while curr:
            sequence.append(str(curr.val))
            curr = curr.prev
        return ' <-> '.join(sequence)


# Example usage
a = Deque()
a.addFirst(1)
a.addFirst(2)
a.addFirst(3)
a.addLast(53)
print(a)

for element in a.Iterator():
    print(element)

File: Assignment 2 - Deques and Randomized Queues/Java/Deque.java
/**
 * The `Deque` class represents a double-ended queue (deque) that allows adding and removing items
 * from both the front and the back. It uses a doubly linked list implementation.
 *
 * @param <Item> the type of items stored in the deque
 */

import edu.princeton.cs.algs4.StdOut;
import java.util.Iterator;

public class Deque<Item> implements Iterable<Item> {
    private class Node {
        Item item;
        Node next;
        Node prev;
    }

    private Node first, last; // Pointers to the first and last nodes
    private int size; // Number of items in the deque

    /**
     * Constructs an empty deque.
     */
    public Deque() {
        first = null;
        last = null;
    }

    /**
     * Checks if the deque is empty.
     *
     * @return true if the deque is empty, false otherwise
     */
    public boolean isEmpty() {
        return first == null && last == null;
    }

    /**
     * Returns the number of items in the deque.
     *
     * @return the number of items
     */
    public int size() {
        return size;
    }

    /**
     * Adds an item to the front of the deque.
     *
     * @param item the item to be added
     * @throws IllegalArgumentException if the input item is null
     */
    public void addFirst(Item item) {
        if (item == null) throw new IllegalArgumentException();
        size++;
        Node newNode = new Node();
        newNode.item = item;

        if (isEmpty()) {
            first = newNode;
            last = newNode;
        }
        else {
            newNode.next = first;
            first.prev = newNode;
            first = newNode;
        }
    }

    /**
     * Adds an item to the back of the deque.
     *
     * @param item the item to be added
     * @throws IllegalArgumentException if the input item is null
     */
    public void addLast(Item item) {
        if (item == null) throw new IllegalArgumentException();
        size++;
        Node newNode = new Node();
        newNode.item = item;

        if (isEmpty()) {
            first = newNode;
            last = newNode;
        }
        else {
            newNode.prev = last;
            last.next = newNode;
            last = newNode;
        }
    }

    /**
     * Removes and returns the item from the front of the deque.
     *
     * @return the removed item
     * @throws java.util.NoSuchElementException if the deque is empty
     */
    public Item removeFirst() {
        if (isEmpty()) {
            throw new java.util.NoSuchElementException();
        }
        size--;
        Item item = first.item;
        first = first.next;
        if (first == null) {
            last = null;
        }
        else {
            first.prev = null;
        }
        return item;
    }

    /**
     * Removes and returns the item from the back of the deque.
     *
     * @return the removed item
     * @throws java.util.NoSuchElementException if the deque is empty
     */
    public Item removeLast() {
        if (isEmpty()) {
            throw new java.util.NoSuchElementException();
        }
        size--;
        Item item = last.item;
        last = last.prev;
        if (last == null) {
            first = null;
        }
        else {
            last.next = null;
        }
        return item;
    }

    /**
     * Returns an iterator over items in order from front to back.
     *
     * @return an iterator
     */
    public Iterator<Item> iterator() {
        return new ListIterator();
    }

    // Private inner class for the iterator
    private class ListIterator implements Iterator<Item> {
        private Node current = first;

        public boolean hasNext() {
            return current != null;
        }

        public void remove() {
            throw new UnsupportedOperationException();
        }

        public Item next() {
            if (!hasNext()) throw new java.util.NoSuchElementException();
            Item item = current.item;
            current = current.next;
            return item;
        }
    }

    // Unit Testing
    public static void main(String[] args) {
        Deque<Integer> deque = new Deque<>();
        deque.addFirst(2);
        deque.addFirst(1);
        deque.addLast(3);
        deque.addLast(4);
        deque.addLast(5);
        for (int i : deque) {
            StdOut.println("Item: " + i);
        }
        StdOut.println("Remove the first item: " + deque.removeFirst());
        StdOut.println("Remove the last item: " + deque.removeLast());
        StdOut.println("Deque is empty? " + deque.isEmpty());
        StdOut.println("Size = " + deque.size());
    }
}
